doc/rst/technotes/optimization.rst

```diff
@@ -0,0 +1,662 @@
+  cases need to be easy to repeat. Usually, that means setting aside some
+  problem configuration and input data to use as the main performance
+  test case. In some cases, you will need a battery of performance
+  tests instead of a single test case.
+
+Timing Regions
+  A common practice is to use :chpl:record:`~Time.stopwatch` to time
+  different regions of your program.  This coarse-grained timing
+  information indicates where your program is spending the most time. The
+  longest-running regions are the places where optimization will give the
+  most improvement to the program's performance. You can print out the
+  performance of each region after it is timed. It's a good idea to use a
+  ``config const`` to make the timing only print when you are studying
+  performance.
+
+Counting Communication Events
+  If you need to do most of your work by running locally oversubscribed
+  (see above) then it's particularly useful to measure performance by
+  counting the number of communication events. Counting communication
+  events can also be useful as a secondary performance metric for
+  multi-node runs. Reducing the number of communication events is a useful
+  strategy for optimizing communication.
+
+  You can use the :chpl:mod:`CommDiagnostics` module and
+  ``startCommDiagnostics()`` followed by ``getCommDiagnostics()`` to
+  count communication events. It's common to
+  ``writeln(getCommDiagnostics()``.  Note that the communication count
+  information will be easier to understand if you compile with
+  ``--no-cache-remote``.
+
+  Communication counts provide a way to compare communication
+  performance independent of where you are running. For example, you
+  might measure and seek to reduce the communication counts when running
+  oversubscribed on a laptop or workstation. Reductions in communication
+  counts from such an effort should also help with multi-node performance
+  when you are able to run on a big system.
+
+
+Overall Flow
+~~~~~~~~~~~~
+
+The general flow of performance optimization will be:
+
+ * use techniques like timing the different phases of your computation to
+   try to understand which parts are the slow or not getting faster as you
+   add more locales or cores
+ * come up with a theory as to what could be improved by thinking about
+   what is happening when your program is running and/or by looking at
+   timing or communication count information
+ * make an adjustment (the adjustment only needs to be good enough to
+   measure its impact on performance at this step)
+ * measure its impact on performance
+ * if it is good, clean up the adjustment, perform correctness testing,
+   and commit/PR it to your program's revision control; if not, try
+   something else
+ * repeat
+
+Configuration Matters
+---------------------
+
+The first thing to check is that you are using Chapel in its
+highest-performing configuration for your system.
+
+ * use ``--fast`` when compiling Chapel programs for performance
```

Comment:
. "The default configuration has additional correctness checks that may worsen performance", or something?

Reply:
I see you mention this below. For this and my `--cache-remote` comment, consider mentioning that the rationale is explained below.